sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java

```diff
@@ -122,7 +122,7 @@ public RelNode convert(RelNode rel) {
    * Returns true only if the literal can be correctly implemented by {@link
    * org.apache.beam.sdk.extensions.sql.impl.rel.BeamCalcRel} in ZetaSQL.
    */
-  private static boolean udfSupportsLiteralType(RelDataType type) {
+  static boolean udfSupportsLiteralType(RelDataType type) {
```

Comment:
It appears `BeamJavaUdfCalcRule`/`BeamZetaSqlCalcRule` are out of sync with their respective `CalcRelSplitter.RelType`. The logic is mostly duplicated between the two, but it is slightly different. Can you make a abstract rule that takes a single `CalcRelSplitter.RelType` and replace the old CalcRules with that?

Reply:
I agree we can find a better way to share all the logic between the two.

Reply:
Done.